### PR TITLE
Research/main parallel tx revision

### PR DESCRIPTION
**Pull request by julianlen**

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java [hunk 1]
```diff
@@ -461,26 +450,23 @@ private BlockResult executeParallel(
...
-                    deletedAccounts,
+                    new HashSet<>(),
                     discardInvalidTxs,
                     acceptInvalidTransactions,
-                    receipts,
-                    executedTransactions,
-                    transactionResults,
+                    new HashMap<>(),
+                    new HashMap<>(),
+                    new HashMap<>(),
```
Comment (javicid) [inline rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java:463]: I would remove these from the constructor as they break encapsulation and can cause problems in the future. Even if it means that we have to manually collect and merge the structures of the last execution. It's just another loop iteration at the end.

File: rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java [hunk 2]
```diff
@@ -549,20 +558,65 @@ private BlockResult executeParallel(
...
         profiler.stop(metric);
         logger.trace("End executeParallel.");
         return result;
     }
 
-    private void addFeesToRemasc(LongAccumulator remascFees, Repository track) {
-        long fee = remascFees.get();
-        if (fee > 0) {
-            track.addBalance(PrecompiledContracts.REMASC_ADDR, Coin.valueOf(fee));
+    private boolean detectCollisions(ReadWrittenKeysTracker tracker) {
```
Comment (javicid) [inline rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java:571]: Shouldn't this method be inside the tracker?

File: rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java [hunk 3]
```diff
@@ -549,20 +558,65 @@ private BlockResult executeParallel(
...
+        threads.addAll(readKeysByThread.keySet());
+        threads.addAll(writtenKeysByThread.keySet());
+
+        for (Long threadId : threads) {
+
+            Set<ByteArrayWrapper> temporalReadKeys = readKeysByThread.getOrDefault(threadId, new HashSet<>());
+            Set<ByteArrayWrapper> temporalWrittenKeys = writtenKeysByThread.getOrDefault(threadId, new HashSet<>());
+
+            if (readKeysSize == 0 && writtenKeysSize == 0) {
+                readKeys = temporalReadKeys;
```
Comment (javicid) [inline rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java:590]: Be careful with this as it copies by reference.

File: rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java [hunk 4]
```diff
@@ -549,20 +558,65 @@ private BlockResult executeParallel(
...
+                readKeysSize = temporalReadKeys.size();
+                writtenKeys = temporalWrittenKeys;
+                writtenKeysSize = temporalWrittenKeys.size();
+                continue;
+            }
+
+            writtenKeys.removeAll(temporalReadKeys);
+            writtenKeys.removeAll(temporalWrittenKeys);
```
Comment (javicid) [inline rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java:598]: Use `Collections.disjoint` better and you can use two nested loops to compare all threads 1v1

**Review by javicid — CHANGES_REQUESTED** [review]

File: rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java (hunk 1 above)
Comment (julianlen) [reply rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java:463]: The last execution needs the merged state. Many of the transactions within the sequential sublist depend on the final state of the parallel execution.

**Pull request merged**